### DISCUSSION TITLE
fetching data

### DISCUSSION THREAD

**Discussion by Adel-abdelfattah** [vercel/next.js#38303] [Help]
I 'm coming from react redux toolkit , I always  use create asyncthunk to fetch data, now there is getServerSideProps , what's the difference between them?

**Comment by icyJoseph** [comment]
Well,

- `getServerSideProps` runs on the server
- If you arrive to the site for the first time, on a page that uses `getServerSideProps`, then the page is built on the server, and the HTML, JS, and CSS that runs it is all sent to the client at once
- If you are already on your site, and navigate to a page that uses `getServerSideProps`,  then the client makes a request for the result of `getServerSideProps`, which arrives as JSON, and it is used as the top level props for the page.

`getServerSideProps`(GSSP) is not necessarily a replacement for your thunks. You can still use your `RTK` fetching methods. 

The advantage of `GSSP` is that it can run on demand, on an isolated environment such as the server. Meaning that you can directly connect to 3rd parties, databases, etc., from `GSSP`, using Node.js APIs and even secrets that you wouldn't want to expose to the server.

It is however costly, because the server has to finish generating the `JSON` or the entire page on demand, every time. 

It is worth understanding that the resulting props from `GSSP` are literally the input to your `page` component. 

It is possible to re-trigger `GSSP` and get fresh props, but this is not necessarily a replacement for a global data provider, such as Redux. The data fetching you'd do with Redux however, happens client side, and often, some might say, a little too late, even if triggered as soon as possible using `useEffects`. With `GSSP` this data is available from the get go! 

> You should use getServerSideProps only if you need to render a page whose data must be fetched at request time. This could be due to the nature of the data or properties of the request (such as authorization headers or geo location). Pages using getServerSideProps will be server side rendered at request time and only be cached if [cache-control headers are configured](https://nextjs.org/docs/going-to-production#caching).
> 
> If you do not need to render the data during the request, then you should consider fetching data on the [client side](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#fetching-data-on-the-client-side) or [getStaticProps](https://nextjs.org/docs/basic-features/data-fetching/get-static-props).


You can see this little demo I put together a while ago, https://github.com/icyJoseph/next-js-react-gbg/blob/main/demo-app/src/pages/collection.tsx, where the collection page uses `GSSP` to calculate from a cookie, the props to pass onto the collection page component. 

With RTK perhaps I'd have to render an empty shell of the page, dispatch and make an async call, to get data from an endpoint and then dispatch that to the store, to finally show the result. In this case with `GSSP` it's all there from the first rendered frame.

### Code Elimination

Last but not least, https://next-code-elimination.vercel.app/, code and packages used only within `GSSP` are removed from the client bundle! So Node.js APIs not supported by the browser, or SECRETS for APIs won't end up on your client bundle.
